api/src/main/java/bio/terra/tanagra/service/query/QueryService.java

```diff
@@ -0,0 +1,44 @@
+package bio.terra.tanagra.service.query;
+
+import bio.terra.tanagra.service.search.Query;
+import bio.terra.tanagra.service.search.SearchContext;
+import bio.terra.tanagra.service.search.Selection;
+import bio.terra.tanagra.service.search.SqlVisitor;
+import bio.terra.tanagra.service.underlay.Underlay;
+import bio.terra.tanagra.service.underlay.UnderlayService;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.util.Optional;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/** A service for executing queries. */
+@Component
+public class QueryService {
+
+  private final UnderlayService underlayService;
+
+  @Autowired
+  public QueryService(UnderlayService underlayService) {
+    this.underlayService = underlayService;
+  }
+
+  /** Generate an SQL query to select the primary ids for the entity of the entity filter. */
+  public String generatePrimaryKeySql(EntityFilter entityFilter) {
+    Optional<Underlay> underlay =
+        underlayService.getUnderlay(entityFilter.primaryEntity().entity().underlay());
+    Preconditions.checkArgument(
+        underlay.isPresent(),
+        "Unable to find underlay '%s'",
+        entityFilter.primaryEntity().entity().underlay());
+
+    Query query =
+        Query.create(
+            ImmutableList.of(
+                Selection.PrimaryKey.create(entityFilter.primaryEntity(), Optional.empty())),
+            entityFilter.primaryEntity(),
+            Optional.of(entityFilter.filter()));
+    return new SqlVisitor(SearchContext.builder().underlay(underlay.get()).build())
```

Comment:
I think you're parsing the parens wrong in this particular case - the SearchContext.Builder is being built, not the `underlay`.
Empirically, taking a builder argument isn't allowed. What is allowed is exposing the builder for a parameter. e.g. `Query.Builder b; b.selectionListBuilder().add(mySelection)'.
We could add our own overloads to take arbitrary arguments, such as builders, but I would probably nit pick it in code review because it makes it ambiguous whether that method modifies the incoming builder in some way.